feat: setup securityContext on plugins

nerzhul opened the pull request:
At Veepee we only permit pods to run as non root. Plugin works properly as non root, add variables to setup it as non root and permit to customize UID

## Description

Ref #121 
## Checklist
- [x] I've read the [guidelines for contributing](https://github.com/aquasecurity/trivy-operator/blob/main/CONTRIBUTING.md) to this repository.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy-operator/tree/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aquasecurity/trivy-operator?pullRequest=362) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aquasecurity/trivy-operator?pullRequest=362) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aquasecurity/trivy-operator?pullRequest=362) it.</sub>

nerzhul commented:
I just fixed tests, it's now mergeable, if you can review and release when you can, we'd glad to do more test on our 200 nodes cluster :)

erikgb commented on docs/operator/configuration.md:
```diff
@@ -50,6 +50,7 @@ To change the target namespace from all namespaces to the `default` namespace ed
 | `scanJob.tolerations`| N/A| JSON representation of the [tolerations] to be applied to the scanner pods so that they can run on nodes with matching taints. Example: `'[{"key":"key1", "operator":"Equal", "value":"value1", "effect":"NoSchedule"}]'`|
 | `scanJob.annotations`| N/A| One-line comma-separated representation of the annotations which the user wants the scanner pods to be annotated with. Example: `foo=bar,env=stage` will annotate the scanner pods with the annotations `foo: bar` and `env: stage` |
 | `scanJob.templateLabel`| N/A| One-line comma-separated representation of the template labels which the user wants the scanner pods to be labeled with. Example: `foo=bar,env=stage` will labeled the scanner pods with the labels `foo: bar` and `env: stage`|
+| `scanJob.podSecurityContext`| N/A| One-line JSON representation of the template securityContext which the user wants the scanner pods to be secured with. Example: `{"RunAsUser": 1000, "RunAsGroup": 1000, "RunAsNonRoot": true}`|
```
Will only "One-line JSON representation" be supported by this? I think it would be nice to support YAML-format - if possible, as this will be a better UX IMO. But probably not possible?

erikgb replied:
Isn't this configured from a configmap? But I see that the similar property `scanJob.templateLabel` (also a typo; should be Label**s**) is using a simple string. I'll leave this decision up to the maintainers....

nerzhul replied:
it's the configmap key exact, i use JSON because GetScanJobTolerations uses also a JSON to be consistent

erikgb commented on pkg/vulnerabilityreport/builder_test.go:
```diff
@@ -207,7 +209,9 @@ func TestScanJobBuilder(t *testing.T) {
 					},
-					Spec: corev1.PodSpec{},
+					Spec: corev1.PodSpec{
+						SecurityContext: &corev1.PodSecurityContext{},
```
I don't think this test should change, if I don't misunderstand the test. As this new feature should not change the existing job spec, if not using this feature. Multiple occurrences.

nerzhul replied:
ok you prefer to keep previous behaviour, and not add any empty spec, let's look at this

erikgb commented on pkg/trivyoperator/config.go:
```diff
@@ -56,6 +56,7 @@ const (
 	keyConfigAuditReportsScanner         = "configAuditReports.scanner"
 	keyScanJobTolerations                = "scanJob.tolerations"
 	keyScanJobAnnotations                = "scanJob.annotations"
+	keyScanJobPodSecurityContext         = "scanJob.podSecurityContext"
```
```suggestion
	keyScanJobPodSecurityContext         = "scanJob.podTemplateSecurityContext"
```

To be better aligned with the existing property `scanJob.podTemplateLabels`?

erikgb commented:
@nerzhul Thanks for this PR! I have added a few comments. We run on Openshift, and it is really picky about securityContext. It has it's SCC that want to control all that.

nerzhul commented:
@erikgb i addressed your concerns, except the one on the format, let's wait for maintainers, but it kept consistency with tolerations, both are JSON unmarshaled.

For Openshift, you have specific things. On our side we use PSP and PodSecurities on all namespaces, then it's not possible to run root pods. With this patch pods runs perfectly as UID > 1000 users and creates reports, on a vanilla Kubernetes with just PSP/PS

erikgb commented:
> For Openshift, you have specific things. 

I am not against adding this new feature, but I don't think it should break the existing operator. :-)

nerzhul commented:
no problem @erikgb , now it's more close to the original one, regarding the tests on other parts :) 
Tests on our dev cluster shows it's fine.
I just amened the helm chart to make helm chart use YAML syntax and let it transform to JSON like tolerations

erikgb commented on pkg/trivyoperator/config.go:
```diff
@@ -118,6 +119,20 @@ func (c ConfigData) GetScanJobTolerations() ([]corev1.Toleration, error) {
+	}
+
+	scanJobPodSecurityContext := &corev1.PodSecurityContext{}
+	err := json.Unmarshal([]byte(c[keyScanJobPodSecurityContext]), &scanJobPodSecurityContext)
```
```suggestion
	err := json.Unmarshal([]byte(c[keyScanJobPodSecurityContext]), scanJobPodSecurityContext)
```

nerzhul replied:
fixed :) ty

nerzhul commented:
Okay i hope we can merge this before you change the whole architecture :) who are maintainers ? in the changelog it seems you are 2 currently active people

josedonizetti commented:
@nerzhul Thanks for the PR, didn't have time to look/test this yet, but I think your change here is addressing what was asked here? https://github.com/aquasecurity/trivy-operator/issues/121

chen-keinan commented:
@nerzhul thanks you for contributing , I'm find with keeping the `podSecurityContext` in json format to be consist with this type of data.
I'll approve it later after I'll test it and make sure it works find.

chen-keinan commented:
I assume it will solve the most cases where it is required in pod level however not if it is required for a specific init container

josedonizetti reviewed (commented):
Looks good! But I do have a comment/question. Considering trivy will scan and report misconfiguration for pods not having a security context with `runAsUser`,  `runAsGroup` and  `runAsNonRoot`. 

https://avd.aquasec.com/misconfig/kubernetes/general/avd-ksv-0012/
https://avd.aquasec.com/misconfig/kubernetes/general/avd-ksv-0020/
https://avd.aquasec.com/misconfig/kubernetes/general/avd-ksv-0021/

Why don't we have a default value for `podTemplateSecurityContext` considering the AVDs above?

josedonizetti reviewed (approved):
I'll create an issue to discuss further if we should have default or not. :)

nerzhul commented:
@josedonizetti it's a very good comment, i think we should have best security options by default, i just kept compat with existing when writing this pr :)

erikgb commented:
> @josedonizetti it's a very good comment, i think we should have best security options by default, i just kept compat with existing when writing this pr :)

It will be a breaking change for users on Openshift, but I do not oppose adding a default according to K8s security best practices. In this case, I think Openshift can improve. 😉 But it should be mentioned in the release notes as a breaking change IMO.
